statistics/handle/handle.go

```diff
@@ -647,64 +647,128 @@ func (h *Handle) LoadNeededHistograms() (err error) {
+	tbl, ok = oldCache.Get(col.TableID)
+	if !ok {
+		return nil
+	}
+	tbl = tbl.Copy()
+	tbl.Columns[c.ID] = colHist
+	if h.updateStatsCache(oldCache.update([]*statistics.Table{tbl}, nil, oldCache.version)) {
+		statistics.HistogramNeededItems.Delete(col)
+	}
+	return nil
+}
+
+func (h *Handle) loadNeededIndexHistograms(reader *statsReader, idx model.TableItemID, loadFMSketch bool) (err error) {
+	oldCache := h.statsCache.Load().(statsCache)
+	tbl, ok := oldCache.Get(idx.TableID)
+	if !ok {
+		return nil
+	}
+	index, ok := tbl.Indices[idx.ID]
+	if !ok {
+		statistics.HistogramNeededItems.Delete(idx)
+		return nil
+	}
+	hg, err := h.histogramFromStorage(reader, idx.TableID, index.ID, types.NewFieldType(mysql.TypeBlob), index.Histogram.NDV, 1, index.LastUpdateVersion, index.NullCount, index.TotColSize, index.Correlation)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cms, topN, err := h.cmSketchAndTopNFromStorage(reader, idx.TableID, 1, idx.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var fms *statistics.FMSketch
+	if loadFMSketch {
+		fms, err = h.fmSketchFromStorage(reader, idx.TableID, 1, idx.ID)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if len(rows) == 0 {
-			logutil.BgLogger().Error("fail to get stats version for this histogram", zap.Int64("table_id", col.TableID), zap.Int64("hist_id", col.ColumnID))
-		}
-		colHist := &statistics.Column{
-			PhysicalID:      col.TableID,
-			Histogram:       *hg,
-			Info:            c.Info,
-			CMSketch:        cms,
-			TopN:            topN,
-			FMSketch:        fms,
-			IsHandle:        c.IsHandle,
-			StatsVer:        rows[0].GetInt64(0),
-			ColLoadedStatus: statistics.NewColFullLoadStatus(),
-		}
-		// Column.Count is calculated by Column.TotalRowCount(). Hence we don't set Column.Count when initializing colHist.
-		colHist.Count = int64(colHist.TotalRowCount())
-		// Reload the latest stats cache, otherwise the `updateStatsCache` may fail with high probability, because functions
-		// like `GetPartitionStats` called in `fmSketchFromStorage` would have modified the stats cache already.
-		oldCache = h.statsCache.Load().(statsCache)
-		tbl, ok = oldCache.Get(col.TableID)
-		if !ok {
-			continue
-		}
-		tbl = tbl.Copy()
-		tbl.Columns[c.ID] = colHist
-		if h.updateStatsCache(oldCache.update([]*statistics.Table{tbl}, nil, oldCache.version)) {
-			statistics.HistogramNeededColumns.Delete(col)
-		}
+	}
+	rows, _, err := reader.read("select stats_ver from mysql.stats_histograms where is_index = 1 and table_id = %? and hist_id = %?", idx.TableID, idx.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		logutil.BgLogger().Error("fail to get stats version for this histogram", zap.Int64("table_id", idx.TableID), zap.Int64("hist_id", idx.ID))
+	}
+	idxHist := &statistics.Index{Histogram: *hg, CMSketch: cms, TopN: topN, FMSketch: fms,
+		Info: index.Info, ErrorRate: index.ErrorRate, StatsVer: index.StatsVer, Flag: index.Flag,
```

Comment:
In `loadNeededColumnHistograms` we use the newly read stats ver(`rows[0].GetInt64(0)`) while we use `index.StatsVer` here.

Reply:
We construct new column/index stats by combining some stats(ndv, nullCount, LastUpdateVersion) in memory and some stats read from disk(buckets, cmsketch, topn). Is it possible to just call `columnStatsFromStorage`/`indexStatsFromStorage` here?

Reply:
I think we can't use `columnStatsFromStorage` or `indexStatsFromStorage` here as they load all column or indice stats while we only need one here.

Reply:
`columnStatsFromStorage`/`indexStatsFromStorage` is only responsible for one column/index. However, in a second look, they may need some modifications to be applied here. Thus not reusing them is ok for me.